### ISSUE TITLE
Differentiable Dual Contouring with GPU QEF Solver

### ISSUE THREAD

**Issue by KhoiDOO** [KhoiDOO/conquer3d#10]
Feature request & tracking for Differentiable Dual Contouring with GPU Quadratic Error Function (QEF) solver for sharp feature and CAD crease preservation.

**Comment by KhoiDOO** [comment]
### Differentiable Dual Contouring with GPU QEF Solver (Usage & Documentation)

#### 1. API Usage & Executable Code Example
```python
import torch
from conquer3d.data_structure import create_voxel_grid, TriangleMesh, create_voxel_grid_from_tmesh
from conquer3d.ops import dual_contouring, dc

# 1. Create 3D Voxel Grid (Dense or Sparse with pad=1)
grid_vertices, voxels, _ = create_voxel_grid(
    grid_min=[-1.0, -1.0, -1.0],
    grid_max=[1.0, 1.0, 1.0],
    res=[64, 64, 64],
    device="cuda"
)

# 2. Evaluate scalar SDF values (and optional vertex colors)
sdf = (torch.norm(grid_vertices, dim=-1) - 0.6).requires_grad_(True)
colors = torch.rand((grid_vertices.shape[0], 3), device="cuda", requires_grad=True)

# 3. Extract Triangle Mesh with Delaunay Max-Min Angle Splitting
verts_tri, faces_tri, out_colors = dc(
    grid_vertices, voxels, sdf, colors=colors, iso=0.0, quad_split=True
)
print(f"Triangles: {verts_tri.shape[0]:,} vertices, {faces_tri.shape[0]:,} faces")

# 4. Extract Pure Quadrilateral Mesh
verts_quad, faces_quad = dc(
    grid_vertices, voxels, sdf, iso=0.0, quad_split=False
)
print(f"Quads: {verts_quad.shape[0]:,} vertices, {faces_quad.shape[0]:,} quads (shape: {faces_quad.shape})")

# 5. Differentiable Autograd Backward Pass
loss = verts_tri.sum() + out_colors.sum()
loss.backward()
print(f"SDF Gradient Norm: {sdf.grad.norm().item():.4f}")
```

#### 2. Key Advantages & Capabilities
- **Sharp Feature Preservation**: Solves the Quadratic Error Function (QEF) using Jacobi $3 \times 3$ symmetric eigensolver SVD, placing dual vertices directly on sharp creases, apexes, and mechanical corners.
- **Pure Quad & Triangle Output**: `quad_split=False` generates direct $(Q, 4)$ quad meshes; `quad_split=True` applies optimal Delaunay max-min angle splitting into $(F, 3)$ triangles.
- **High Throughput CUDA Kernels**: Achieves $>230\text{ M faces/s}$ on modern NVIDIA GPUs (RTX 4090).
- **Fully Differentiable**: Backpropagates analytical gradients directly into `sdf` values and `colors` features.
- **100% Watertight Manifold Topology**: Exact Cartesian quadrant indexing eliminates self-intersecting bowtie quads, yielding $\chi=2$ closed 2-manifold surfaces.

#### 3. Disadvantages & Known Trade-offs
- **Single Dual Vertex per Cell**: Like standard Dual Contouring (Ju et al. 2002), each voxel cell contains at most one dual vertex. Complex sub-voxel topology with multiple disconnected sheets inside a single voxel cell can share vertices (pinch points).
- **Narrow-Band Boundary Padding**: On sparse voxel grids from BVH ray/AABB tests, `pad=1` is required so that all 4 incident voxels surrounding every sign-crossing edge are present in GPU memory.
- **Normal Dependence**: For razor-sharp knife edges on CAD models, passing exact Hermite/triangle normals via `grid_normals` avoids trilinear gradient smoothing across $C^0$ slope discontinuities.

#### 4. Empirical Benchmarks & Performance Metrics (NVIDIA GeForce RTX 4090)
| Model / Test | Resolution | Extracted Vertices | Extracted Faces | Extraction Latency | Throughput | Watertight |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| **Sphere SDF** | $32^3$ | 1,610 | 3,216 triangles (1,608 quads) | 0.82 ms | $3.9\text{ M faces/s}$ | **Yes** ($\chi=2$) |
| **Sharp Box SDF** | $32^3$ | 1,538 | 3,072 triangles (1,536 quads) | 0.79 ms | $3.8\text{ M faces/s}$ | **Yes** (Exact 0.4000) |
| **Armadillo** | $64^3$ | 6,021 | 12,080 triangles (6,040 quads) | 4.19 ms | $2.8\text{ M faces/s}$ | **Yes** |
| **Fandisk CAD** | $1024^3$ (Sparse) | **1,716,386** | **3,432,768 triangles** | **14.36 ms** | **$239\text{ M faces/s}$** | **Yes** (0 holes) |
| **Fandisk CAD (Quads)** | $1024^3$ (Sparse) | **1,716,386** | **1,716,384 quads** | **12.81 ms** | **$134\text{ M quads/s}$** | **Yes** |

#### 5. Technical Insights & Architectural Decisions
- **Deterministic Cartesian Quadrant Slotting**: Moving from floating-point polar angle sorting (`atan2f`) to a static compile-time lookup table `dc_edge_quadrant[12]` completely eliminated self-intersecting quads caused by coincident dual vertices on sharp features.
- **Regularized SVD Truncation**: Singular values $\sigma_i < 0.01 \sigma_{\max}$ are clamped/truncated with mass-centroid regularization, preventing dual vertices from drifting outside the voxel boundary when normals are nearly coplanar.

**Comment by KhoiDOO** [comment]
Implemented and covered by PR.

**Comment by KhoiDOO** [comment]
The backward function has not been tested.
